### PR TITLE
WMFC-69: Forgot Username

### PR DESCRIPTION
**Pull request by eduardo-villarreal**

### CODE REVIEW

File: simplesamlphp/modules/core/templates/forgot_username.tpl.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?php
+$this->data['header'] = $this->t('{login:user_pass_header}');
+
+(strlen($this->data['username']) > 0) ? $this->data['autofocus'] = 'password' : $this->data['autofocus'] = 'username';
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/modules/core/templates/forgot_username.tpl.php:4]: Please change this shape, ternary operation must be all right to the assignment part

File: simplesamlphp/modules/core/templates/forgot_username.tpl.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+$this->includeAtTemplateBase('includes/header-fullpage.php');
+
+?>
+<div id="fullpage-wrap">
+    <div id="fullpage-header">
+        <?php
+            echo '<a id="fullpage-header-arrow-back" href="javascript:history.back()"><span class="material-icons">arrow_back_ios</span></a>';
+        ?>
+        <div id="fullpage-header-title">Retrieve User Name</div>
+        <div></div>
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/modules/core/templates/forgot_username.tpl.php:15]: If this is to put some space at the bottom please user CSS instead. 
Note: do not use inline styles

File: simplesamlphp/modules/core/www/forgot_username.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    \SimpleSAML\Utils\Auth::requireAdmin();
+}
+$loginurl = \SimpleSAML\Utils\Auth::getAdminLoginURL();
+$isadmin = \SimpleSAML\Utils\Auth::isAdmin();
+$logouturl = \SimpleSAML\Utils\Auth::getAdminLogoutURL();
+
+$links = [];
+$links_welcome = [];
+$links_config = [];
+$links_auth = [];
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/modules/core/www/forgot_username.php:18]: What's the difference between
Line 18: $links_auth = [];
and
Line 21: $allLinks = [

They both look to be the same if so, use one single assignment statement for it

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
+    </div><!-- #content -->
+        </div><!-- #wrap -->
+        <script>
+            let emailInput = document.getElementById('forgot-username-email-input');
+            let emailError = document.getElementById('forgot-username-email-error');
+            let notification = document.getElementById('forgot-username-notification');
+            let sendLoader = document.getElementById('forgot-username-submit-loader');
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:7]: Please check if you can use **const** instead

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+            function checkNotEmpty(value) {
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:9]: This method looks unneeded, a valid/invalid validation over JS string can be done with 

if (stringValue) or even if (!!stringValue) if you want explicit boolean conversion

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                errors.forEach(({err, input, msg}) => {
+                    if (!checkNotEmpty(input.value)) {
+                        err.style.color = 'rgba(209,91,93,1)';
+                        input.style.borderColor = 'rgba(203,70,73,1)';
+                        err.innerText = `The ${msg} field is required.`;
+                    } else {
+                        err.style.color = 'rgba(255,255,255,1)';
+                        input.style.borderColor = 'rgba(221,221,221,1)';
+                        err.innerText = '';
+                    }
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:27]: Use CSS Classes instead we should not have explicit style manipulation in code files

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                if (/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/.test(emailInput.value)) {
+                    emailError.style.color = 'rgba(255,255,255,1)';
+                    emailInput.style.borderColor = 'rgba(221,221,221,1)';
+                    emailError.innerText = '';
+                    return true;
+                } else {
+                    emailError.style.color = 'rgba(209,91,93,1)';
+                    emailInput.style.borderColor = 'rgba(203,70,73,1)';
+                    emailError.innerText = `The Email Address field must be a valid email address.`;
+                    return false;
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:41]: Use CSS Classes instead we should not have explicit style manipulation in code files

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            }
+
+            function sendData() {
+                let validationType = 1;
+
+                sendLoader.style.display = 'block';
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:48]: Use CSS Classes instead we should not have explicit style manipulation in code files

File: simplesamlphp/templates/includes/footer-forgot-username.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                        }).catch((e) => {
+                            sendLoader.style.display = 'none';
+                            notification.style.width = '90%';
+                            notification.style.borderRadius = '3px';
+                            notification.style.backgroundColor = 'rgba(243,217,218,1)';
+                            notification.style.color = 'rgba(105,36,38,1)';
+                            notification.style.padding = '20px';
+                            notification.style.margin = '0 0 10px 0';
+                            notification.style.border = '1px solid rgba(238,200,203,1)';
+                            notification.innerText = 'An error occurred, please try again later.';
```
Comment (angel-moreno-wedbush) [inline simplesamlphp/templates/includes/footer-forgot-username.php:87]: Use CSS Classes instead we should not have explicit style manipulation in code files

File: simplesamlphp/modules/core/templates/forgot_username.tpl.php (hunk 2 above)
Comment (eduardo-villarreal) [reply simplesamlphp/modules/core/templates/forgot_username.tpl.php:15]: There are no inline styles, this is for flex to grab 3 items and center the arrow and the title.

File: simplesamlphp/modules/core/www/forgot_username.php (hunk 3 above)
Comment (eduardo-villarreal) [reply simplesamlphp/modules/core/www/forgot_username.php:18]: This is how SimpleSAMLPhp is managing the pages links. (This comes by default)

File: simplesamlphp/templates/includes/footer-forgot-username.php (hunk 5 above)
Comment (eduardo-villarreal) [reply simplesamlphp/templates/includes/footer-forgot-username.php:9]: Reusing code so the files look alike. Will change.

**Pull request merged**